glue/sales_val_flagging.py

```diff
@@ -152,30 +156,115 @@ def group_size_adjustment(df, stat_groups: list, min_threshold, condos: bool):
     )
 
-    # Modify the .loc condition to include checking for sv_outlier_type values
-    condition = (merged_df["_merge"] == "both") & (merged_df["sv_is_outlier"] == 1)
+    # Assign sv_outlier_reasons
+    for idx in range(1, 4):
+        df[f"sv_outlier_reason{idx}"] = np.nan
+
+    outlier_type_crosswalk = {
+        "sv_ind_price_high_price": "High price",
+        "sv_ind_ptax_flag_w_high_price": "High price",
+        "sv_ind_price_low_price": "Low price",
+        "sv_ind_ptax_flag_w_low_price": "Low price",
+        "sv_ind_price_high_price_sqft": "High price per square foot",
+        "sv_ind_ptax_flag_w_high_price_sqft": "High price per square foot",
+        "sv_ind_price_low_price_sqft": "Low price per square foot",
+        "sv_ind_ptax_flag_w_low_price_sqft": "Low price per square foot",
+        "sv_ind_ptax_flag": "PTAX-203 Exclusion",
+        "sv_ind_char_short_term_owner": "Short-term owner",
+        "sv_ind_char_family_sale": "Family Sale",
+        "sv_ind_char_non_person_sale": "Non-person sale",
+        "sv_ind_char_statistical_anomaly": "Statistical Anomaly",
+        "sv_ind_char_price_swing_homeflip": "Price swing / Home flip",
+    }
+
+    """
+    During our statistical flagging process, we automatically discard
+    a sale's eligibility for outlier status if the number of sales in 
+    the statistical grouping is below a certain threshold. The list - 
+    `group_thresh_price_fix` along with the ['_merge'] column will allow
+    us to exclude these sales for the sv_is_outlier status.
+
+    Since the `sv_is_outlier` column requires a price value, we simply
+    do not assign these price outlier flags if the group number is below a certain
+    threshold
+
+    Note: This doesn't apply for sales that also have a ptax outlier status.
+          In this case, we still assign the price outlier status.
+    """
+    group_thresh_price_fix = [
+        "sv_ind_price_high_price",
+        "sv_ind_price_low_price",
+        "sv_ind_price_high_price_sqft",
+        "sv_ind_price_low_price_sqft",
+    ]
 
-    # Using .loc[] to set the desired values for rows meeting the condition
-    merged_df.loc[condition, "sv_outlier_type"] = "Not outlier"
-    merged_df.loc[condition, "sv_is_outlier"] = 0
+    def fill_outlier_reasons(row):
+        reason_idx = 1
+        reasons_added = set()  # Set to track reasons already added
+
+        for reason in outlier_type_crosswalk:
+            current_reason = outlier_type_crosswalk[reason]
+            # Add a check to ensure that only specific reasons are added if _merge is not 'both'
+            if (
+                reason in row
+                and row[reason]
+                and current_reason
+                not in reasons_added  # Check if the reason is already added
+                # Apply group threshold logic
+                and not (row["_merge"] == "both" and reason in group_thresh_price_fix)
```

Comment:
**[Question, non-blocking]** I don't totally get how this line works, can you break it down for me?

Reply:
### Condition 1  
`reason_ind_col in row`: this checks if the column exists in the dataset. This check is necessary because the sqft columns won't exist in the condos data

### Condition 2
`and row[reason_ind_col]`: this is the check for whether the ind col is true

### Condition 3

This checks to make sure that the column doesn't already exist in the `reasons_added_set`. Given that it is a set, it is kind of a redundant operation, but does it make sense to keep this for computation reduction? 

```python
and current_reason
not in reasons_added
```

Also I think it acts as a good check for `row[f"sv_outlier_reason{len(reasons_added) + 1}"] = current_reason` to make sure the column doesn't get populated with an incorrect reason name.


### Condition 4

`and not (row["_merge"] == "both" and reason_ind_col in group_thresh_price_fix`

This condition makes sure a price value won't be assigned to an obs with a group of N < `group_thresh`

Reply:
Thanks for these responses!

> [Condition 3] Given that it is a set, it is kind of a redundant operation, but does it make sense to keep this for computation reduction?

I think it makes sense to keep it -- the set will automatically deduplicate indicator values, but we still want to make sure we don't add duplicate values to `row[f"sv_outlier_reason{len(reasons_added) + 1}"]`.

> [Condition 4] This condition makes sure a price value won't be assigned to an obs with a group of `N < group_thresh`

Got it, that makes sense! Do you mind if we adjust the comment to make this a little bit clearer?

```python
            if (
                reason_ind_col in row
                and row[reason_ind_col]
                and current_reason
                not in reasons_added  # Check if the reason is already added
                # Apply group threshold logic: `row["_merge"]` will be `both` when the group threshold
                # is met, but only price indicators (`group_thresh_price_fix`) should use this threshold,
                # since ptax indicators have thresholds built-in
                and not (row["_merge"] == "both" and reason_ind_col in group_thresh_price_fix)
            ):
```